homeassistant/components/ring/siren.py

```diff
@@ -1,21 +1,83 @@
 """Component providing HA Siren support for Ring Chimes."""
 
+from collections.abc import Callable, Coroutine
+from dataclasses import dataclass
 import logging
-from typing import Any
+from typing import Any, Generic, cast
 
-from ring_doorbell import RingChime, RingEventKind
+from ring_doorbell import (
+    RingCapability,
+    RingChime,
+    RingEventKind,
+    RingGeneric,
+    RingStickUpCam,
+)
 
-from homeassistant.components.siren import ATTR_TONE, SirenEntity, SirenEntityFeature
-from homeassistant.core import HomeAssistant
+from homeassistant.components.siren import (
+    ATTR_TONE,
+    SirenEntity,
+    SirenEntityDescription,
+    SirenEntityFeature,
+    SirenTurnOnServiceParameters,
+)
+from homeassistant.const import Platform
+from homeassistant.core import HomeAssistant, callback
 from homeassistant.helpers.entity_platform import AddEntitiesCallback
 
 from . import RingConfigEntry
 from .coordinator import RingDataCoordinator
-from .entity import RingEntity, exception_wrap
+from .entity import (
+    RingDeviceT,
+    RingEntity,
+    RingEntityDescription,
+    async_check_create_deprecated,
+    refresh_after,
+)
 
 _LOGGER = logging.getLogger(__name__)
 
 
+@dataclass(frozen=True, kw_only=True)
+class RingSirenEntityDescription(
+    SirenEntityDescription, RingEntityDescription, Generic[RingDeviceT]
+):
+    """Describes a Ring siren entity."""
+
+    exists_fn: Callable[[RingGeneric], bool]
+    unique_id_fn: Callable[[RingDeviceT], str] = lambda device: str(
+        device.device_api_id
+    )
+    is_on_fn: Callable[[RingDeviceT], bool] | None = None
+    turn_on_fn: (
+        Callable[[RingDeviceT, SirenTurnOnServiceParameters], Coroutine[Any, Any, Any]]
+        | None
+    ) = None
+    turn_off_fn: Callable[[RingDeviceT], Coroutine[Any, Any, None]] | None = None
```

Comment:
why are these nullable? `turn_on` is at least set everytime

Reply:
I’m attempting to have an entity description that follows the logic of the `SirenEntity` for which TURN_ON is a feature that may not be supported.